### DISCUSSION TITLE
Novide displays different colors in `toggleterm` than Neovim on the Terminal app.

### DISCUSSION THREAD

**Discussion by azzamsa** [neovide/neovide#2252] [Q&A] [answered]
Hi, 👋🏼

Novide displays different colors in `toggleterm` than Neovim on the Terminal app.

- Neovide. 

The blue suggestions are hard to read

![image](https://github.com/neovide/neovide/assets/17734314/e17faefa-3586-477a-924b-a3b24d2ace94)

- Neovim in Terminal.

It works great for me.

![image](https://github.com/neovide/neovide/assets/17734314/4f8159cc-de68-47ae-bdaa-b3f170e45c86)

**Comment by fredizzimo** [comment]
https://neovide.dev/faq.html#the-terminal-displays-fallback-colorsterminal-does-not-show-my-colors

Some color schemes also need some option to enable the terminal colors.

**Reply by azzamsa** [reply]
In the screenshot above, the values of `vim.g.terminal_color_N` have been set in [this section](https://github.com/folke/tokyonight.nvim/blob/f247ee700b569ed43f39320413a13ba9b0aef0db/lua/tokyonight/util.lua#L126) of the Tokyonight plugin. Tokyonight defaults to `terminal_colors = true`, which automatically configures these values.

I also tried setting `terminal_colors = false`, and I can confirm that Tokyonight correctly adjusts these values. Otherwise, my Neovim terminal would look terrible, as shown in the screenshot below.

![image](https://github.com/neovide/neovide/assets/17734314/df90b262-fa6a-4d80-81de-bc7e2eba46c2)


I also attempted to manually set them by changing every blue color to another color, but I still see the blue color in Neovide's terminal.

```lua
on_colors = function(colors)
    if vim.g.neovide then
        -- dark
        vim.g.terminal_color_0 = colors.black
        vim.g.terminal_color_8 = colors.terminal_black

        -- light
        vim.g.terminal_color_7 = colors.fg_dark
        vim.g.terminal_color_15 = colors.fg

        -- colors
        vim.g.terminal_color_1 = colors.red
        vim.g.terminal_color_9 = colors.red

        vim.g.terminal_color_2 = colors.green
        vim.g.terminal_color_10 = colors.green

        vim.g.terminal_color_3 = colors.yellow
        vim.g.terminal_color_11 = colors.yellow

        vim.g.terminal_color_4 = colors.green
        vim.g.terminal_color_12 = colors.green

        vim.g.terminal_color_5 = colors.magenta
        vim.g.terminal_color_13 = colors.magenta

        vim.g.terminal_color_6 = colors.cyan
        vim.g.terminal_color_14 = colors.cyan
    end
end,
```

**Reply by fredizzimo** [reply]
Are you on macOS? The lack of srgb rendering might explain it https://github.com/neovide/neovide/issues/2044

**Reply by azzamsa** [reply]
No, I am on Linux.

```
NVIM v0.9.4
neovide 0.12.1
```

```
- **OS Name:**                                     Fedora Linux 39 (Workstation Edition)
- **OS Type:**                                     64-bit
- **GNOME Version:**                               45.2
- **Windowing System:**                            Wayland
- **Kernel Version:**                              Linux 6.6.9-200.fc39.x86_64
```

**Reply by azzamsa** [reply]
I have set every color in `g:terminal_color_<0-14>`  to green. but the blue is still there.

In Emacs, I have describe-face at point. Is there a way in Neovide/Nvim where  I can track where the color comes from?

**Reply by fredizzimo** [reply]
Hm, maybe something else in your config is overwriting it? Can you echo the value of the colors to verify, `:echo g:terminal_color_0` for example, to verify?

If you run `neovide --log` it's possible to see what all colors are defined to, and what's used for rendering

**Reply by azzamsa** [reply]
Got it. I have the correct color when running Neovide from the terminal.

![image](https://github.com/neovide/neovide/assets/17734314/98a51eaf-e045-4f2a-a618-c4d8365eeef1)

But, not when I am running Neovide from GUI (GNOME app menu).

![image](https://github.com/neovide/neovide/assets/17734314/d2fb044b-4d09-4ac6-bfe8-b9d5a2d918ef)

I don't think this is the case: https://neovide.dev/faq.html#neovide-is-not-picking-up-some-shell-configured-information, because all other env I set in my shell are loaded correctly.

Any idea why?

**Reply by fredizzimo** [reply]
It probably is because of that, the shell probably has it's own theming configuration and uses true colors, so unaffected by the terminal colors, that is not run.

**Reply by azzamsa** [reply]
I have a similar situation with Flameshot. The solution is to use a custom script to launch the app and point the `Exec` to the script.

I tried for Neovide, but no luck. The blue is still there.
Do you have any other solution?

- neovide.desktop

```
[Desktop Entry]
Type=Application
Exec=neoviden
```

- neoviden

```bash
#!/bin/bash

neovide
```

**Reply by fredizzimo** [reply]
Which shell are you using? I think it can fixed by changing the location of the theming, so that it's loaded, when Neovide starts the terminal. I'm not sure why it isn't right now, but if you can find out where it's done, we can figure it out.

**Reply by azzamsa** [reply]
Got it!

I am using Fish.
I tried to change the value below, from gray (#636da6) to green (#c3e88d).

```fish
set -l duck c3e88d #c3e88d # suggestion, comment

set -g fish_color_autosuggestion $duck
set -g fish_color_comment $duck
```

So the behavior (bug?) is that Neovide uses a brighter color than Nvim (Terminal).
It is *why*  my previous gray color turns blue-ish in Neovide.

Here. When I set it to soft green (#c3e88d), Neovide use very bright green.

![image](https://github.com/neovide/neovide/assets/17734314/a569ac88-7c66-4872-aaa5-867ee7ddc5e3)

Neovim (in Teminal) uses correct color

![image](https://github.com/neovide/neovide/assets/17734314/6804de61-7fbf-4e5b-b4f3-391e2fb3c0f9)

The color should look like this.

![image](https://github.com/neovide/neovide/assets/17734314/d95c432a-7e2b-4995-879b-a231cb91d262)

---

Is this a bug?

**Reply by fredizzimo** [reply]
I think this is some form of srgb/color space issue, but it's strange that the other colors are correc, or maybe they are not?

I'm an ipad right now, and don't have a color picker available, but could you check what values it gives for the two greens?

**Reply by fredizzimo** [reply]
The resulting color in Neovide appears to be #cccc66, I have no idea why.

Could you provide a log file? That should give more information, what kind of colors we get from Neovim.

**Reply by azzamsa** [reply]
Neovide log. Executed with `neovide --log` on Terminal app.

[neovide_rCURRENT.log](https://github.com/neovide/neovide/files/13872465/neovide_rCURRENT.log)

**Reply by fredizzimo** [reply]
I will check the log soon, but I think I know what's going on. It's using libvterms default 256 color palette. https://github.com/neovim/neovim/issues/7018

I don't know if fish can be forced to true color, but that would likely solve the issue.

**Reply by azzamsa** [reply]
- Neovide from Terminal

![image](https://github.com/neovide/neovide/assets/17734314/00be1f42-d529-4338-9e78-1a99a3ed160d)

- Neovide from GNOME apps

![image](https://github.com/neovide/neovide/assets/17734314/d1002b72-afb4-436d-9a57-032703c398ac)

- But Neovide from GNOME apps able to display color correctly outside floating terminal/toggle term

![image](https://github.com/neovide/neovide/assets/17734314/ffd0c721-ba4f-4915-8e4d-9f0bc038b57c)

**Reply by fredizzimo** [reply]
Could you run this https://raw.githubusercontent.com/roosta/colors/master/scripts/support/256-colors.sh

Linked from here  https://github.com/neovim/neovim/issues/7018

And check if the color you get is from the palette?

**Reply by azzamsa** [reply]
> And check if the color you get is from the palette?

Yes, I am able to get the palette right

![image](https://github.com/neovide/neovide/assets/17734314/06075e01-3b8d-4327-b334-be57f7f74fe3)

**Reply by fredizzimo** [reply]
Then it's even more weird, if the built in terminal actually can display that color.

**Reply by fredizzimo** [reply]
Ah wait no, that's one of the first 16 colors.

Do you find the wrong cccc66 color in the palette?

**Reply by azzamsa** [reply]
> Do you find the wrong cccc66 color in the palette?

Yes.

![image](https://github.com/neovide/neovide/assets/17734314/6569178c-dd66-45b9-8285-25dbac8a2671)


Sorry for a late reply 🙏🏼

**Reply by fredizzimo** [reply]
That's most likely what happens then. The neovim libvterm finds the closest matching palette entry and uses that, with it's hardcoded values. There does not seem to be any way around that.

But if you run it from the terminal it uses the palette of the terminal instead, as described in the issue I linked earlier.

What's your $TERM variable inside floatterm, it might be possible to use true colors instead of a palette by changing that.

**Reply by fredizzimo** [reply, marked as answer]
Yes `set -g fish_term24bit 1`

https://fishshell.com/docs/current/cmds/set_color.html

**Reply by fredizzimo** [reply]
A more proper fix is probably for Neovide to set `term_name` and `term_colors` https://neovim.io/doc/user/ui.html#ui-ext-options

**Reply by azzamsa** [reply]
> What's your $TERM variable inside floatterm?

```
🦄 echo $TERM
xterm-256color
```

> Yes `set -g fish_term24bit 1`

I tried to toggle this value on and off, and yes. it this the step that solves my issue

![image](https://github.com/neovide/neovide/assets/17734314/42b3217b-34d4-461b-9c16-17afbc17c8e7)


---

I can't believe that this is not a Neovide issue, but you've accompanied me this far. 
Thank you so much, Fredizzimo. Thank you! 🎖️

**Reply by azzamsa** [reply]
![image](https://github.com/neovide/neovide/assets/17734314/f9e4bc82-cea0-41a3-87f0-23135f01a320)

Thank you, fredizzimo! 🎖️

**Reply by fredizzimo** [reply]
There's actually an even better solution 

https://neovim.io/doc/user/options.html#'termguicolors'

I found out by reading this https://github.com/termstandard/colors, so setting `COLORTERM=truecolor` is the standard way to control this, and should work both for fish and many other applications.

And setting `termguicolors` controls that variable for the builtin terminal.

We should probably set it by default in Neovide.
